dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/ListSink.kt

```diff
@@ -0,0 +1,202 @@
+        destEnd: Int,
+        isNull: BooleanArray,
+    ) {
+        while (data.size < destBegin) {
+            _data += null
+            hasNulls = true
+        }
+        for ((srcIndex, _) in (destBegin..<destEnd).withIndex()) {
+            _data += getValue(src, srcIndex, isNull)
+        }
+    }
+
+    private fun writeReplacing(
+        src: Any,
+        destBegin: Int,
+        destEnd: Int,
+        isNull: BooleanArray,
+    ) {
+        for ((srcIndex, destIndex) in (destBegin..<destEnd).withIndex()) {
+            _data[destIndex] = getValue(src, srcIndex, isNull)
+        }
+    }
+
+    override fun write(
+        src: Any,
+        isNull: BooleanArray,
+        destBegin: Long,
+        destEnd: Long,
+        appending: Boolean,
+    ) {
+        if (destBegin == destEnd) return
+        val destBeginAsInt = destBegin.toInt()
+        val destEndAsInt = destEnd.toInt()
+        if (appending) {
+            writeAppending(src = src, destBegin = destBeginAsInt, destEnd = destEndAsInt, isNull = isNull)
+        } else {
+            writeReplacing(src = src, destBegin = destBeginAsInt, destEnd = destEndAsInt, isNull = isNull)
+        }
+    }
+
+    override fun read(
+        dest: Any,
+        isNull: BooleanArray,
+        srcBegin: Long,
+        srcEnd: Long,
+    ) {
+        if (srcBegin == srcEnd) return
+        val srcBeginAsInt = srcBegin.toInt()
+        val srcEndAsInt = srcEnd.toInt()
+
+        when (dataType) {
+            BYTE -> {
+                dest as ByteArray
+                for ((srcIndex, destIndex) in (srcBeginAsInt..<srcEndAsInt).withIndex()) {
+                    val value = data[srcIndex] as Byte?
+                    if (value != null) dest[destIndex] = value
+                    isNull[destIndex] = value == null
+                }
+            }
+
+            SHORT -> {
+                dest as ShortArray
+                for ((srcIndex, destIndex) in (srcBeginAsInt..<srcEndAsInt).withIndex()) {
+                    val value = data[srcIndex] as Short?
+                    if (value != null) dest[destIndex] = value
+                    isNull[destIndex] = value == null
+                }
+            }
+
+            INT -> {
+                dest as IntArray
+                for ((srcIndex, destIndex) in (srcBeginAsInt..<srcEndAsInt).withIndex()) {
+                    val value = data[srcIndex] as Int?
+                    if (value != null) dest[destIndex] = value
+                    isNull[destIndex] = value == null
+                }
+            }
+
+            LONG -> {
+                dest as LongArray
+                for ((srcIndex, destIndex) in (srcBeginAsInt..<srcEndAsInt).withIndex()) {
+                    val value = data[srcIndex] as Long?
+                    if (value != null) dest[destIndex] = value
+                    isNull[destIndex] = value == null
+                }
+            }
+
+            else -> error("unsupported sink state")
```

Comment:
Could you please improve the error message! Or give some advice

Reply:
The datatype supplied here can only be byte, short, int, or long, as described by https://github.com/deephaven/deephaven-csv/blob/main/ADVANCED.md#supporting-the-fast-path-for-numeric-type-inference, so this case never occurs, unless deephaven updates and does something weird. But I'll add a comment